node_versions.go

```diff
@@ -0,0 +1,190 @@
+package iavl
+
+import (
+	"fmt"
+)
+
+// not Thread-Safe
+type NodeVersions struct {
+	nums    []int
+	changes map[int64]int // version -> num, changes will be merged to nums when commit
+
+	maxVersions int
+	maxNodes    int
+
+	firstVersion   int64
+	nextVersion    int64
+	nextVersionIdx int
+
+	totalNodes int
+}
+
+func NewNodeVersions(maxVersions int, maxNodes int, lastVersion int64) *NodeVersions {
+	return &NodeVersions{
+		nums:        make([]int, maxVersions, maxVersions),
+		changes:     make(map[int64]int, 32),
+		maxVersions: maxVersions,
+		maxNodes:    maxNodes,
+
+		firstVersion:   lastVersion - 1,
+		nextVersion:    lastVersion,
+		nextVersionIdx: 0,
+		totalNodes:     0,
+	}
+}
+
+func (nv *NodeVersions) Inc1(version int64) {
+	nv.changes[version]++
+}
+
+func (nv *NodeVersions) Inc(version int64, times int) {
+	nv.changes[version] += times
+}
+
+func (nv *NodeVersions) Dec1(version int64) {
+	nv.changes[version]--
```

Comment:
should we check nv.changes[version] > 0 before we minus one

Reply:
no, the changes can be negative